### ISSUE TITLE
oneOf/anyOf doesn't seem to work with objects

### ISSUE THREAD

**Issue by bluenote10** [fastify/fast-json-stringify#290]
## 🐛 Bug Report

oneOf/anyOf doesn't seem to work with objects.

## To Reproduce

The following schema models a union type, i.e. in TypeScript terms:

```ts
type Foo = { foo: string, baz: string };
type Bar = { bar: string, baz: string };
type FooBar = Foo | Bar;
```

```js
const fastJson = require("fast-json-stringify");

const schemaFoo = {
  properties: {
    foo: { type: "string" },
    baz: { type: "string" },
  },
  required: ["foo", "baz"],
};

const schemaBar = {
  properties: {
    bar: { type: "string" },
    baz: { type: "string" },
  },
  required: ["bar", "baz"],
};

const stringify = fastJson({
  type: "object",
  anyOf: [schemaFoo, schemaBar],
});

console.log(stringify({ foo: "foo", baz: "baz" }));
console.log(stringify({ bar: "bar", baz: "baz" }));
```

Output:

```
{}
{}
```

## Expected behavior

Since both objects match the schema, both should be stringified.

## Your Environment

- *node version*: v14.16.0
- *fastify version*: 3.13.0
- *fast-json-stringify*: 2.4.2
- *os*: Ubuntu 18.04

**Comment by mcollina** [comment]
Would you mind trying to add `type: 'object'` to both schemas?

**Comment by bluenote10** [comment]
Yes I did that as well before, but it didn't make a difference.

**Comment by Eomm** [comment]
this schema works:

```
const stringify = fastJson({
  anyOf: [schemaFoo, schemaBar]
})
```

Defining the `type` beside the `anyOf` would conflict when your data could be an object or a string like:

```
const stringify = fastJson({
  anyOf: [schemaFoo, schemaBar, { type: 'string' }]
})
```

**Comment by bluenote10** [comment]
Good catch, it makes certainly more sense without specifying `type` in the top level schema. I'm not sure where I got that from. So perhaps there is no need to support that.

I just did a quick check and ran

```json
{
  "type": "object",
  "anyOf": [
    {
      "properties": {
        "foo": { "type": "string" }
      }
    },
    {
      "properties": {
        "bar": { "type": "string" }
      }
    }
  ]
}
```

through the [online JSON schema validator](https://www.jsonschemavalidator.net/), which considers it valid against all schema drafts, but probably cannot verify such semantic violations anyway.

**Comment by mcollina** [comment]
I think this is a bug in this module: we should throw if we cannot deal with this ambiguity. Better of, we could handle it if it wasn't really hard.

The schema is valid JSON schema (we validate it on creation).

**Comment by mcollina** [comment]
Would you like to send a PR?

**Comment by bluenote10** [comment]
Sure, I can have a look, but it may take a bit until I get to it.

**Comment by stalniy** [comment]
same issue on my side, trying to serialize

```ts
const schema = object({
    messages: array({
        anyOf: [
          object({
            id: string,
            createdAt: date,
            author: object({
              _id: string,
              fullName: string,
            }),
            type: { type: "string", enum: ["Text"] },
            body: string,
          }),
          object({
            id: string,
            createdAt: date,
            author: object({
              _id: string,
              fullName: string,
            }),
            type: { type: "string", enum: ["StatusTransition"] },
            status: values(VacancyApplicationStatusEnum),
          }),
        ]
      })
})
```

And I get :

```json
{
  "messages": [
    null
  ]
}
```

**Comment by stalniy** [comment]
is there any support for discriminated union types?

**Comment by Eomm** [comment]
> is there any support for discriminated union types?

As a general rule, I suggest running the [debugMode](https://github.com/fastify/fast-json-stringify#debug-mode) whenever there is some issue to understand what is going on.

Debugging a bit your gist with this feature I found:

- if you change the schema removing the `type`, the serializer function output would be fine.
this is a bug in FJS and it would be great if you would like to fix it (ignore the type when there is the `anyOf` keyword)

```diff
  items: {
-    type: "object",
    anyOf: [
```

- looking deeply at the schema compared to the input object, the `id`s field must be `string`s instead of `integer`

With these two changes, your code will works 👍🏽

**Comment by Eomm** [comment]
> discriminator keyword? 

No: it is a OpenAPI keyword and fast-json-stringify requires a JSON Schema Draft 7

**Comment by michael-melodyarc** [comment]
> > discriminator keyword?
> 
> No: it is a OpenAPI keyword and fast-json-stringify requires a JSON Schema Draft 7

Sorry, to clarify, it's not that I need support for the `discriminator` keyword specifically, but rather that keyword uses `oneOf`, and since the `discriminator` keyword requires `type` set at the top level, it hits this bug where it simply returns an empty object in serialization.  Bumping the type down into the `oneOf` subschemas fixes it, but then `ajv` throws warnings/errors by default since the top-level `type` keyword is missing.

**Comment by gfortaine** [comment]
@stalniy `oneOf` should to the trick to ensure [type coercion](https://ajv.js.org/coercion.html) 👍 :

```javascript
const compile = require("fast-json-stringify");

const schemaFoo = {
  type: "object",
  required: ["id", "createdAt", "author", "type", "body"],
  properties: {
    id: { type: "string" },
    createdAt: { type: "string", format: "date-time" },
    author: {
      type: "object",
      required: ["_id", "fullName"],
      properties: { _id: { type: "string" }, fullName: { type: "string" } },
      additionalProperties: false,
    },
    type: { type: "string", enum: ["Text"] },
    body: { type: "string" },
  },
  additionalProperties: false,
};

const schemaBar = {
  type: "object",
  required: ["id", "createdAt", "author", "type", "status"],
  properties: {
    id: { type: "string" },
    createdAt: { type: "string", format: "date-time" },
    author: {
      type: "object",
      required: ["_id", "fullName"],
      properties: { _id: { type: "string" }, fullName: { type: "string" } },
      additionalProperties: false,
    },
    type: { type: "string", enum: ["StatusTransition"] },
    status: {
      type: "string",
    },
  },
  additionalProperties: false,
};

const schema = {
  type: "array",
  items: {
    oneOf: [schemaFoo, schemaBar],
  },
};

const stringify = compile(schema);

console.log(
  stringify([
    {
      type: "StatusTransition",
      id: "1",
      createdAt: new Date().toISOString(),
      status: "bla",
      author: {
        _id: "2",
        fullName: "John Doe",
      },
    },
  ])
);

**Comment by gfortaine** [comment]
@stalniy Here is another implementation based on [AJV JTD Serializer ](https://ajv.js.org/guide/getting-started.html#parsing-and-serializing-json) that should ensure better soundness regarding the validation of the schema cc @epoberezkin

```javascript
const Ajv = require("ajv/dist/jtd");

const ajv = new Ajv(); // options can be passed, e.g. {allErrors: true}

const schemaFoo = {
  properties: {
    id: { type: "string" },
    createdAt: { type: "timestamp" },
    author: {
      properties: { _id: { type: "string" }, fullName: { type: "string" } },
    },
    type: { enum: ["Test"] },
    body: { type: "string" },
  },
};

const schemaBar = {
  properties: {
    id: { type: "string" },
    createdAt: { type: "timestamp" },
    author: {
      properties: { _id: { type: "string" }, fullName: { type: "string" } },
    },
    type: { enum: ["StatusTransition"] },
    status: {
      type: "string",
    },
  },
};

const schema = {
  elements: {
    metadata: {
      union: [schemaFoo, schemaBar],
    },
  },
};

// const dataFoo = [
//   {
//     type: "Test",
//     body: "bla",
//     id: "1",
//     createdAt: new Date(),
//     author: {
//       _id: "2",
//       fullName: "John Doe",
//     },
//   },
// ];

const dataBar = [
  {
    type: "StatusTransition",
    status: "bla",
    id: "1",
    createdAt: new Date(),
    author: {
      _id: "2",
      fullName: "John Doe",
    },
  },
];

const validate = ajv.compile(schema);
const serialize = ajv.compileSerializer(schema);

function validateAndSerialize(data) {
  const valid = validate(data);

  if (valid) {
    console.log(serialize(data));
  } else {
    console.log("invalid", validate.errors);
  }
}

for (const data of [
  // dataFoo,
  dataBar,
]) {
  validateAndSerialize(data);
}

**Comment by thom4parisot** [comment]
I seem to have ran in this issue with a GeoJSON structure in which I had a `oneOf` of Polygon or MultiPolygon schema on a `geometry` property.

It scrambled the body of the request, altering the first coordinates of the geometry.
Shall I add more info here or rather open a new issue?

**Comment by StrahinjaDucic** [comment]
Hi everybody, is this problem fixed or not? Because I still have similar issue with anyOf. 
const responseSchema = {
    207: {
        type: 'array',
        items: {
            anyOf: [successfullyDuplicatedFile, unsuccessfullyDuplicatedFile],
        },
        minItems: 1,
    }
} as const;

const successfullyDuplicatedFile = {
    type: 'object',
    properties: {
        id: fileId,
        statusCode: { type: 'number' },
        response: {
            type: 'object',
            properties: {
                ...someSchema,
            },
            description: 'Duplicated file.',
        },
    },
    required: ['id', 'statusCode', 'response'],
} as const;

const unsuccessfullyDuplicatedFile = {
    type: 'object',
    properties: {
        id: fileId,
        statusCode: { type: 'number' },
        response: {
            type: 'object',
            properties: {
                message: { type: 'string' },
            },
            required: ['message'],
            description: 'Error message.',
        },
    },
    required: ['id', 'statusCode', 'response'],
} as const;

Can someone help me, thanks !!!

**Comment by Shiva127** [comment]
Any update on this issue?

**Comment by Farenheith** [comment]
Hello, I have a suggestion for this issue as a workaround: when I define a type as `{}` in the schema, I suppose the library falls back to JSON.stringify, creating a proper result. Wouldn't it be a better behavior for oneOf or anyOf than the current one?

I know this introduces a new issue: fields that we didn't want to be exposed in the serialization would start to be there, so this shouldn't go in the library, as whoever defines the schema can just set it like that.

The workaround here could be to print a warning once per schema, perhaps, about how this narrowing cannot be done reliably. Unless we could rely on named types (ie, `instance.constructor === refType`), it's not possible to define a simple way to narrow down types properly here.

A proper fix would create an algorithm to analyse the schema options and "compile" a comparison function to decide between the options based on the input object. I suppose it'd be performant most of the time, but we'd still have some edge cases where it could affect performance (which could go against the purpose of the lib) or would even be impossible to differentiate between them.

Another idea would be to offer a way for the user to provide a comparison function, maybe as an option with a Record of keys as JSON paths, values as those functions?

**Comment by Tony133** [comment]
This has been **fixed**. On `v7.0.1` the schema from the original report serializes correctly:

```js
const stringify = fastJson({ type: 'object', anyOf: [schemaFoo, schemaBar] })

stringify({ foo: 'foo', baz: 'baz' }) // → {"foo":"foo","baz":"baz"}
stringify({ bar: 'bar', baz: 'baz' }) // → {"bar":"bar","baz":"baz"}
```

The variant reported later in the thread (`items: { type: 'object', anyOf: [...] }`) works as well.

Closing as completed. If you still hit this on the latest release, please open a new issue with a minimal reproduction in a git-clonable repository.
